[TOPI] Resize rework

jwfromm opened the pull request:
This PR includes a rework of how topi handles resizing. Rather than have a call into a c++ implementation, it makes sense to have all resizing operations as tensor expressions as this allows easier customization and enables fast scheduling where appropriate (it's also much more readable).

Besides adding a true tvm implementation of nearest_neighbor, bilinear, and bicubic resampling, I include a few minor changes that are important to note (and may be controversial): 

- align_corners now always defaults to True. The false setting literally only exists to support a tensorflow bug that has stayed around for backwards compatibility reasons and is being fully removed in the tf2.0 release. We shouldn't be propagating this bug by default. That said, all methods support align_corners=False when needed.

- I toned down the size of one of the crop_and_resize tests as it was occasionally tripping correctness due to floating point arithmetic errors.

- Although the c++ workflow is completely removed from relay, nnvm still uses it so I haven't completely cut it out of topi. If nnvm weren't deprecating soon I would be more motivated to make this change.

- Bicubic resizing is supported and generates good looking images. However, I haven't found two bicubic implementations that produce the exact same results, tensorflow, PIL, and this implementation all produce visually similar but slightly different results for example. This makes testing a little complicated so it's not currently included.

- Currently resize is always scheduled as injective (just like the previous version of resize), however this may not be optimal. I imagine a more efficient schedule could be created, especially for bicubic resizing which can be quite slow. 

I'm curious to know what the reviewers think of these above points and can address them if needed.

tqchen commented:
@masahi @kazum can you help to manage this PR?

masahi commented:
ok, I'm somewhat familiar with resize issues so I can take a look. Also cc @srkreddy1238

masahi commented:
Regarding c++ vs python: I think it is fine to move operator implementations to python if the operator is "sufficiently complex". 

There was a time (before AutoTVM and Relay) when people tried to make dependencies on python smaller and ported many topi python ops to c++. Before that upsampling and resize were written in python. But some operators like convolution will never be ported to c++ anyway and AutoTVM only works with python implementation.

masahi commented on topi/python/topi/image/resize.py:
```diff
@@ -32,18 +35,178 @@ def resize(data, size, layout="NCHW", align_corners=False, method="BILINEAR"):
+        To preserve the values at the corner pixels.
 
-    method: {"BILINEAR", "NEAREST_NEIGHBOR"}
+    method: {"bilinear", "nearest_neighbor", "bicubic"}
```
+1 for this. To me all caps look weird.

masahi commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -481,7 +481,8 @@ def global_avg_pool2d(data,
                layout="NCHW",
-               method="NEAREST_NEIGHBOR"):
+               method="NEAREST_NEIGHBOR",
+               align_corners=True):
```
Since nearest is the default for upsampling, I think it makes more sence if the default for align_corners is False.

jwfromm replied:
I agree and have changed the defaults for upsampling.

masahi commented:
@jwfromm I think making align_corners default to True is good for resize, but for upsampling it should default to False. I believe nearest is the most common use case for upsampling, and in that case align_corners can be ignored.

jwfromm commented:
I think that's a good point, @masahi . I've changed the default behavior of upsampling to use align_corners=False as you recommend (and updated testing accordingly). It's interesting to hear that resize/upsampling were written in python at some point but moved to C++. Are there any big upsides to moving away from the python implementation? I was under the impression an implementation like this would be an upgrade.

masahi commented on topi/include/topi/nn/upsampling.h:
```diff
@@ -55,7 +55,7 @@ inline Tensor upsampling(const Tensor& input,
                          std::string name = "tensor",
                          std::string tag = kInjective) {
-  return resize(input, shape, layout, false, mode);
+  return resize(input, shape, layout, true, mode);
```
This change should be removed. (I know it is no longer used)

jwfromm replied:
You're right, I've made this change along with reverting some of the changes I made to frontends that are no longer needed due to defaulting upsampling to align_corners=False.

masahi commented:
@srkreddy1238 @FrozenGene @kazum @vinx13 
I'm going to merge this if there is no further comment. align_corners defaults to True for the resize op, this is potentially a breaking change.

vinx13 commented on topi/python/topi/nn/upsampling.py:
```diff
@@ -20,7 +20,7 @@
 
 
-def upsampling(data, scale, layout="NCHW", method='NEAREST_NEIGHBOR'):
+def upsampling(data, scale, layout="NCHW", method='NEAREST_NEIGHBOR', align_corners=False):
```
What about also changing methods here to lower cased

jwfromm replied:
You're right, that's better. Now updated.

FrozenGene commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -504,14 +505,17 @@ def upsampling(data,
         The computed result.
     """
-    return _make.upsampling(data, scale, layout, method)
+    return _make.upsampling(data, scale, layout, method, align_corners)
```
Do we really need to change the name to lower case? We have many places using upper case. Such as here and many frontend (such as TFLite)

FrozenGene commented on include/tvm/relay/attrs/image.h:
```diff
@@ -50,8 +51,11 @@ struct ResizeAttrs : public tvm::AttrsNode<ResizeAttrs> {
         .describe("Specify the mode to use for scaling."
                   "NEAREST_NEIGHBOR -  Nearest Neighbor"
                   "BILINEAR - Bilinear Interpolation");
```
We have supported "BICUBIC" in this PR. I think could expose it in relay and frontend. Again, I think using uppercase is acceptable. Otherwise I think you need change many places (relay, frontend, topi and so on) to unified.

FrozenGene commented on include/tvm/relay/attrs/nn.h:
```diff
@@ -394,6 +395,8 @@ struct UpSamplingAttrs : public tvm::AttrsNode<UpSamplingAttrs> {
         .describe("Specify the mode to use for scaling."
                   "NEAREST_NEIGHBOR -  Nearest Neighbor"
                   "BILINEAR - Bilinear Interpolation");
```
BICUBIC

FrozenGene commented on python/tvm/relay/op/nn/nn.py:
```diff
@@ -504,14 +505,17 @@ def upsampling(data,
 
     method : str, optional
-        Scale method to used [NEAREST_NEIGHBOR, BILINEAR].
+        Scale method to used [nearest_neighbor, bilinear, bicubic].
```
Uppercase or lower case to be discussed as previous.

FrozenGene commented on tests/python/relay/test_op_level2.py:
```diff
@@ -513,12 +513,14 @@ def get_shape():
+                            method=method, align_corners=align_corners)
     func = relay.Function([x], y)
     data = np.random.uniform(size=dshape).astype(dtype)
     if method == "NEAREST_NEIGHBOR":
```
As you see, we use upper case here again now... Too confused, if we change to lower case...

masahi commented:
@FrozenGene The option string is converted to lower case [here](https://github.com/dmlc/tvm/pull/3788/files#diff-909bdfb6ebd66e7b7d102ec945ca2e44R56), so both lower and upper case work. Existing code won't break.

FrozenGene commented:
> @FrozenGene The option string is converted to lower case [here](https://github.com/dmlc/tvm/pull/3788/files#diff-909bdfb6ebd66e7b7d102ec945ca2e44R56), so both lower and upper case work. Existing code won't break.

Hi, thanks for pointing it. However, My main concern is the unified style. Like this:
https://github.com/dmlc/tvm/blob/41b7cb1071fb869db0dd5a98c2875bfbfc291f89/python/tvm/relay/op/nn/nn.py#L484-L508

We accept the upper case, but then we comment the method used is lower case. Which will make others confuse until he read the code `method.lower`.

masahi commented:
If we want to pick either lower or upper case, I prefer converting everything to lower case, and keep `method.tolower`.

FrozenGene commented:
> If we want to pick either lower or upper case, I prefer converting everything to lower case, and keep `method.tolower`. Lower case is common in other frameworks too.

Lower case / upper case both is ok for me. But we should keep one style (frontend, relay, topi).

jwfromm commented:
I think that's fair. I've done a sweep and lower cased everything appropriate. I also added the relay bicubic option to the documentation.

masahi commented:
Thanks @jwfromm @FrozenGene @vinx13 this is merged.
